v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/transforms/SourceWriterFn.java

```diff
@@ -0,0 +1,335 @@
+  private static final Logger LOG = LoggerFactory.getLogger(SourceWriterFn.class);
+  private transient ObjectMapper mapper;
+
+  private final Counter numRecProcessedMetric =
+      Metrics.counter(SourceWriterFn.class, "records_processed");
+
+  private final Distribution lagMetric =
+      Metrics.distribution(SourceWriterFn.class, "replication_lag_in_milli");
+  private transient Map<String, MySqlDao> mySqlDaoMap = new HashMap<>();
+
+  private final Schema schema;
+  private final String sourceDbTimezoneOffset;
+  private final List<Shard> shards;
+  private final SpannerConfig spannerConfig;
+  private transient SpannerDao spannerDao;
+  private final Ddl ddl;
+  private final String shadowTablePrefix;
+
+  public SourceWriterFn(
+      List<Shard> shards,
+      Schema schema,
+      SpannerConfig spannerConfig,
+      String sourceDbTimezoneOffset,
+      Ddl ddl,
+      String shadowTablePrefix) {
+
+    this.schema = schema;
+    this.sourceDbTimezoneOffset = sourceDbTimezoneOffset;
+    this.shards = shards;
+    this.spannerConfig = spannerConfig;
+    this.ddl = ddl;
+    this.shadowTablePrefix = shadowTablePrefix;
+  }
+
+  /** Setup function connects to Cloud Spanner. */
+  @Setup
+  public void setup() {
+    mapper = new ObjectMapper();
+    mapper.enable(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS);
+
+    // TODO: Support multiple databases
+    for (Shard shard : shards) {
+      String sourceConnectionUrl =
+          "jdbc:mysql://" + shard.getHost() + ":" + shard.getPort() + "/" + shard.getDbName();
+      mySqlDaoMap = new HashMap<>();
+      mySqlDaoMap.put(
+          shard.getLogicalShardId(),
+          new MySqlDao(sourceConnectionUrl, shard.getUserName(), shard.getPassword()));
+    }
+    spannerDao = new SpannerDao(spannerConfig);
+  }
+
+  /** Teardown function disconnects from the Cloud Spanner. */
+  @Teardown
+  public void teardown() throws Exception {
+    if (mySqlDaoMap != null) {
+      for (MySqlDao mySqlDao : mySqlDaoMap.values()) {
+        mySqlDao.cleanup();
+      }
+    }
+    spannerDao.close();
+    mySqlDaoMap.clear();
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c) {
+    KV<Long, TrimmedShardedDataChangeRecord> element = c.element();
+    TrimmedShardedDataChangeRecord spannerRec = element.getValue();
+    String shardId = spannerRec.getShard();
+    // Get the latest commit timestamp processed at source
+    try {
+      JsonNode keysJson = mapper.readTree(spannerRec.getMod().getKeysJson());
+      String tableName = spannerRec.getTableName();
+      com.google.cloud.spanner.Key primaryKey = changeEventToPrimaryKey(tableName, keysJson);
+      String shadowTableName = shadowTablePrefix + tableName;
+      boolean isSourceAhead = false;
+
+      com.google.cloud.Timestamp processedCommitTimestamp =
+          spannerDao.getProcessedCommitTimestamp(shadowTableName, primaryKey);
+      isSourceAhead =
+          processedCommitTimestamp != null
```

Comment:
isShourceAhead should be false if processedCommitTimestamp == null also.

Reply:
yes the above condition will set to false in case the processedCommitTimestamp is null